### PR TITLE
make IPv6 compatible, safe run for coroutine interrupting

### PR DESCRIPTION
**Pull request by AllentDan**
#463
#408 ctrl+C
fix api_client.py reuse the old session

### CODE REVIEW

File: lmdeploy/serve/openai/api_server.py [hunk 1]
```diff
@@ -72,6 +71,16 @@ async def check_request(request) -> Optional[JSONResponse]:
     return ret
 
 
+def ip2id(host_ip: str):
+    """Convert host ip address to session id."""
+    if '.' in host_ip:  # IPv4
+        return int(host_ip.replace('.', ''))
+    if ':' in host_ip:  # IPv6
+        return int(host_ip.replace(':', ''))
```
Comment (tpoisonooo) [inline lmdeploy/serve/openai/api_server.py:79]: 这样肯定会溢出啊， 32 个 0xFF char 直接转，肯定不对。

**Review by tpoisonooo — APPROVED** [review]

**Comment by Nick-bit233** [comment]
I ran into this IPv6 problem and it seems can be sovled after I checkout to this pr. So I decide to leave a mark here. Will there be another Reviewer that can make the fix be merged soon? Thanks for your efforts.

File: lmdeploy/serve/openai/api_server.py [hunk 2]
```diff
@@ -296,21 +291,26 @@ async def generate(request: GenerateRequest, raw_request: Request = None):
...
-                'text': out.response,
-                'tokens': out.generate_token_len,
-                'finish_reason': out.finish_reason
-            }
-            yield (json.dumps(ret) + '\0').encode('utf-8')
+            chunk = GenerateResponse(text=out.response,
+                                     tokens=out.generate_token_len,
+                                     finish_reason=out.finish_reason)
+            data = chunk.model_dump_json()
+            yield f'{data}\n'
```
Comment (lvhan028) [inline lmdeploy/serve/openai/api_server.py:298]: Should we encode it to `utf-8`? And why `\n` is appended?

Comment (AllentDan) [reply lmdeploy/serve/openai/api_server.py:298]: `StreamingResponse` will automatically encode it to `utf-8`.
`\n` is just a delimiter for client to iterate over the response data.

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**